src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/vlboptionsfinder/VideoLinkBookingOptionsFinder.kt

```diff
@@ -0,0 +1,158 @@
+package uk.gov.justice.digital.hmpps.whereabouts.services.vlboptionsfinder
+
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.whereabouts.dto.prisonapi.ScheduledAppointmentDto
+import uk.gov.justice.digital.hmpps.whereabouts.services.locationfinder.Interval
+import java.time.LocalTime
+import java.util.TreeMap
+
+sealed class Event(val time: LocalTime)
+class StartEvent(time: LocalTime) : Event(time)
+class EndEvent(time: LocalTime) : Event(time)
+
+/**
+ * Takes a list of Events representing the start and end of appointments during a day.
+ * Uses this information to answer the question 'is the Timeline available (free of appointments) during this interval'
+ * (isFreeForInterval(interval: Interval) below.
+ */
+class Timeline(events: List<Event>) {
+
+  /*
+    The periods of time when there are no appointments.
+    The keys are the times at which an empty period started
+    while the values are the corresponding times when the empty period ended.
+   */
+  private val emptyPeriods = TreeMap<LocalTime, LocalTime>()
+
+  init {
+    // The number of bookings in this room.
+    var currentBookings = 0
+
+    // The time of the previous event
+    var previousEventTime: LocalTime = LocalTime.MIN
+
+    // The time when the room last became unoccupied.
+    // This will be null when the room is currently occupied.
+    var freePeriodStarted: LocalTime? = LocalTime.MIN
+
+    events
+      .plus(StartEvent(LocalTime.MAX))
+      .sortedBy { it.time }
+      .forEach {
+        if (it.time > previousEventTime) {
```

Comment:
This is clearly quite complicated hence much more comments than code but it feels like you could change this to a couple of `when` cases and also move the comments to explain each?

Reply:
Interesting.  I'll try that out.